### ISSUE TITLE
no-implicit-globals Should report an error for `let`/`const`/`class`

### ISSUE THREAD

**Issue by Jamesernator** [eslint/eslint#9537]
Something I noticed when fixing up some code manually (thanks to those who fixed [this](https://github.com/eslint/eslint/issues/9520) which allowed me to start using it and discover this other one) is that there's no warning for top-level `const`/`let`/`class` in the top level using the rule `no-implicit-globals`.

Basically those declarations *do* introduce a global, it's just not attached window/global/self:

```html
<!doctype html>
<script>
    var one = 1;
    let two = 2;
    const three = 3;
    function four() {}
    class Five {}
</script>

<script>
    // Different script but the globals are still shared
    console.log(one); // 1
    console.log(two); // 2
    console.log(three); // 3
    console.log(four); // function four() {}
    console.log(Five); // class Five() {}

    console.log(window.one); // 1
    console.log(window.two); // undefined, let/const/class aren't added to
                            // the global object but are still global
    var one = 12;
    console.log(one); // 12
</script>

<script>
    // But const/let for declared variables are SyntaxErrors that prevent
    // the whole script running
    
    console.log("This doesn't get printed at all!");
    const two = 12; // SyntaxError: Identifier 'two' has already been declared
</script>
```

It would make sense for it to follow the inverse rules as #9525, basically it would only report if neither `ecmaFeatures: { globalReturn: true }` or `sourceType: 'module'` were true.

**Comment by not-an-aardvark** [comment]
I think the technical distinction here is that top-level `var` creates a property on the global object, whereas `let` and `const` declare variables in the [global lexical environment](https://tc39.github.io/ecma262/#global-environment) (i.e. the global scope) without modifying the global object.

That said, it seems unlikely that someone would want to allow one without allowing the other.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
